2-virtual-module-source.md

```diff
@@ -0,0 +1,408 @@
+This would protect the invariant that module sources are serializable.
+
+### Emulated JavaScript
+
+[Should virtual module sources support emulated
+JavaScript?](https://github.com/tc39/proposal-compartments/issues/70)
+That will require, in some cases, separation of initialization from execution
+as separate phases.
+Should the virtual module source protocol support separate paths for modules
+that do not require an initialization phase?
+
+### Shape of the internal namespace object
+
+The `execute` method of a virtual module source needs access to both
+the global object and internal view of bindings.
+These could be addressed singly by a reification of the module environment
+record, or with separate objects as written.
+Future amendments to modules may eventually also add lexical names to the
+module environment record that are not properties of the global object.
+
+## Limitations
+
+Module sources compiled by the `ModuleSource` constructor capture enough data
+that engines can transfer them in many ways.
+The internal representation of a module source can be an immutable record that
+a JavaScript host can trivially share throughout an agent cluster.
+Communicating JavaScript agent clusters can transfer module sources as data.
+
+In the most naive case, a module source is just a record of the original source
+and its type, which is not limited to JavaScript modules, since hosts can define
+additional module source types.
+For example, [`WebAssembly.Module`][web-assembly-module] might be trivially
+extended to serve as a module source, by adding a [[ModuleSource]] slot
+referring to a concrete ***Module Source Record***.
+In this case, it is sufficient for a pair of JavaScript agent clusters
+to transfer the original source and type.
+
+In a more elaborate case, the module source does not even retain the original
+text, but its compiled bytecode.
+In this case, a pair of compatible JavaScript agent clusters might elect to
+send and receive byte code.
+
+Virtualized module sources are transmissible only between JavaScript agent clusters
+if both the sender and receiver agree on a protocol for constructing a new
+virtual module source from some serial representation.
+
+To wit, it would not be possible for a host to transmit arbitrary virtual
+module sources between agents using a general purpose algorithm like
+[structured clone][structured-clone].
+The protocol for communicating virtual module sources between hosts would
+necessarily need to be implemented in user code.
+Any general-purpose host-defined mechanism for transmitting module graphs would
+necessarily fail when encountering a virtual module source.
+For example, it may be possible to extend structured clone to transmit
+instances of `ModuleSource` between any JavaScript hosts, or even
+[`WebAssembly.Module`][web-assembly-module] when a host-defined extension is
+present, but it would not be possible for this behavior to generalize to
+virtual module sources.
```

Comment:
Could we design the virtual module API such that we could actually do so? For example Node's experimental loader API has a mechanism through which to [communicate evaluation between threads](https://nodejs.org/api/esm.html#globalpreload) (as the loader is always run on a separate thread from the main code).

We could do something similar by having a `remoteExecute` hook that returns a `SourceTextModule` (or script?) to evaluate on the remote host e.g. something like:

```js
class JsonModuleSource {
     bindings = [{ export: "default" }];

     #text = text;
     constructor(text) {
         this.#text = text;
         JSON.parse(text); // check parsing is successful
     }

    executeRemote() {
        // Supposing the extension of having module blocks return
        // Module objects we could provide a module block that will be evaluated
        // on whatever thread
        return {
            // structured cloned and sent along with the module
            data: { text: this.#text },
            module: module {
                export function execute(namespace, data) {
                    namespace.default = data.text;
                }
            },
        };
    }
}

const jsonModule = new Module(new JsonModuleSource("{}"));
const worker = new Worker("./worker.js");

// the module gets the executeRemote data cloned so we have everything on the other
// thread in order to be able to execute the module on any thread
worker.postMessage({ jsonModule });
```

Reply:
@Jamesernator I would like to continue this conversation, possibly on another issue.

One of the motivations for separating ***Module Source Record*** from ***Module Record*** is that this will allow us to share module sources between threads. `ModuleSource` instances are thin objects with a [[ModuleSource]] internal slot. So, a thread pool could conceivably fetch, compile, analyze a module graph in parallel and trivially transfer the compiled sources to the main thread for instantiation.

That would work for everything *except* virtual module sources. The underlying text could be fetched on a thread, but it’d have to be transferred and lifted into an instance on the main.

And, the other limitation is that between workers, virtual modules can only be transferred by a user-code protocol that knows about all the virtual module types. That is really just a generalization of the underlying truth though: structured clone would have to know about every kind of module source it can transfer too.